src/rfq/protocols/bebop/client.rs

```diff
@@ -0,0 +1,592 @@
+#![allow(dead_code)] // TODO remove this
+use std::collections::{HashMap, HashSet};
+
+use async_trait::async_trait;
+use futures::{stream::BoxStream, StreamExt};
+use serde::{Deserialize, Serialize};
+use tokio_tungstenite::{connect_async_with_config, tungstenite::Message};
+use tracing;
+
+use crate::{
+    rfq::{
+        client::RFQClient,
+        errors::RFQError,
+        indicatively_priced::SignedQuote,
+        models::{GetAmountOutParams, TimestampHeader},
+    },
+    tycho_client::feed::synchronizer::{ComponentWithState, Snapshot, StateSyncMessage},
+    tycho_common::dto::{ProtocolComponent, ResponseProtocolState},
+    tycho_core::dto::Chain,
+};
+
+type BebopPriceMessage = HashMap<String, BebopPriceData>;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct BebopPriceData {
+    last_update_ts: f64,
+    /// Vec where each tuple is (price, size)
+    bids: Vec<(f64, f64)>,
+    /// Vec where each tuple is (price, size)
+    asks: Vec<(f64, f64)>,
+}
+
+impl BebopPriceData {
+    /// Calculates Total Value Locked (TVL) based on bid/ask levels.
+    ///
+    /// TVL is calculated using the formula from Bebop's documentation:
+    /// https://docs.bebop.xyz/bebop/bebop-api-pmm-rfq/rfq-api-endpoints/pricing#interpreting-price-levels
+    ///
+    /// Returns the average of bid and ask TVLs across all price levels.
+    ///
+    /// Note: This calculation assumes all pairs use the same quote token.
+    /// For cross-pair comparisons, token prices should be normalized to a common denomination.
+    fn calculate_tvl(&self, quote_price_data: Option<BebopPriceData>) -> f64 {
+        let bid_tvl: f64 = self
+            .bids
+            .iter()
+            .map(|(price, size)| price * size)
+            .sum();
+
+        let ask_tvl: f64 = self
+            .asks
+            .iter()
+            .map(|(price, size)| price * size)
+            .sum();
+
+        let mut total_tvl = (bid_tvl + ask_tvl) / 2.0;
+
+        // If quote price data is provided, we need to normalize the tvl
+        if let Some(quote_data) = quote_price_data {
+            if let Some(quote_price_usd) = quote_data.get_mid_usd_price(total_tvl) {
+                total_tvl *= quote_price_usd;
+            } else {
+                // Quote token has no TVL is USD - return 0
+                return 0.0;
+            }
+        }
+        total_tvl
+    }
+
+    /// Gets the mid price of the given token with USDC as the quote token
+    ///
+    /// # Parameters
+    /// - `token_amount`: The amount of tokens to price
+    /// - `price_data`: The price data containing bids and asks
+    ///
+    /// # Returns
+    /// The USDC amount at mid price, or None if insufficient liquidity on either side
+    fn get_mid_usd_price(&self, token_amount: f64) -> Option<f64> {
```

Comment:
I find it strange that we mention usd but we don't even check if it is usd. This method and `calculate_usd_amount` are general, they are not specific to usd. We are just relying on self price levels to have a certain quote token.